tests/0151-purge-brokers.c

```diff
@@ -0,0 +1,410 @@
+}
+
+/**
+ * @brief Test adding and removing brokers from the mock cluster.
+ *        Verify that the client is updated with the new broker list.
+ *
+ *        All \p actions are executed in sequence.
+ *
+ *        After each action, the client is expected to have the broker
+ *        ids in \p expected_broker_ids and the count to be
+ *        \p expected_brokers_cnt .
+ *        Brokers are added at the head of the list.
+ *
+ *        @param initial_cluster_size Initial number of brokers in the cluster.
+ *        @param actions Array of actions to perform. Each action is a pair
+ *                       (action,broker id). 0 to remove, 1 to add,
+ *                       2 to set down, 3 to set up.
+ *        @param expected_broker_ids Array of broker ids expected after each
+ *                                   action.
+ *        @param expected_broker_ids_cnt Number of elements in
+ *                                       \p expected_broker_ids .
+ *        @param expected_brokers_cnt Array of expected broker count after each
+ *                                    action.
+ *        @param await_verification If `rd_false`, the verification is
+ *                                  done only after last action.
+ */
+#define TEST_ACTION_REMOVE_BROKER         0
+#define TEST_ACTION_ADD_BROKER            1
+#define TEST_ACTION_SET_DOWN_BROKER       2
+#define TEST_ACTION_SET_UP_BROKER         3
+#define TEST_ACTION_SET_GROUP_COORDINATOR 4
+#define TEST_GROUP                        "topic1"
+static void
+do_test_add_remove_brokers(int32_t initial_cluster_size,
+                           int32_t actions[][2],
+                           int32_t expected_broker_ids[][5],
+                           size_t expected_broker_ids_cnt,
+                           int32_t expected_brokers_cnt[],
+                           void (*edit_configuration_cb)(rd_kafka_conf_t *conf),
+                           rd_bool_t (*await_after_action_cb)(int action)) {
+        rd_kafka_mock_cluster_t *cluster;
+        const char *bootstraps;
+        rd_kafka_conf_t *conf;
+        rd_kafka_t *rk;
+        size_t action = 0;
+
+        cluster = test_mock_cluster_new(initial_cluster_size, &bootstraps);
+
+        test_conf_init(&conf, NULL, 100);
+
+        test_conf_set(conf, "bootstrap.servers", bootstraps);
+        test_conf_set(conf, "group.id", TEST_GROUP);
+        test_conf_set(conf, "topic.metadata.refresh.interval.ms", "1000");
+        if (edit_configuration_cb) {
+                edit_configuration_cb(conf);
+        }
+
+        rk = test_create_handle(RD_KAFKA_CONSUMER, conf);
+        test_consumer_subscribe(rk, TEST_GROUP);
+
+        /* Create a new topic to trigger partition reassignment */
+        rd_kafka_mock_topic_create(cluster, TEST_GROUP, 3,
+                                   initial_cluster_size);
+
+        for (action = 0; action < expected_broker_ids_cnt; action++) {
+                if (action > 0) {
```

Comment:
Let's remove this condition altogether. Add a new action to do nothing and add first action in the actions list as that action. Improves readability of the test.

Reply:
I don't think it improves readability because there are N actions it's expected there are N+1 states including the initial state. Here it's ensuring that we're in the initial state before executing the first action.

Reply:
I am talking about initial state. This seems like a hardcoding for the initial state. We can skip this if condition if we just add new action to do nothing.